Jeong p

JeongP opened the pull request.

JeongP committed:
각 태그에 id값 부여하기

JeongP committed:
README에 구현해야 할 기능 재정의 하기

JeongP committed:
car 생성중 귀가

JeongP committed:
display 처리를 위한 class -> id 변환

JeongP committed:
RacingGame 생성자를 통한 초기 세팅

JeongP committed:
초기 view 세팅 처리를 위한 모듈화

JeongP committed:
Update README.md

JeongP committed:
수도코딩으로 틀잡기

JeongP committed:
readme pull & merge

JeongP committed:
readme 수행 사항 반영

JeongP committed:
자동차 입력 받기 및 조건 검사 기능 구현

JeongP committed:
시도할 횟수 입력받기 기능 구현

JeongP committed:
게임시작: 자동차 객체 생성 및 race 기능 구현

JeongP committed:
레이스 진행 될 때마다 결과 출력 기능 구현

JeongP committed:
우승자 선정 기능 구현

JeongP committed:
우승자 출력 기능 구현

JeongP committed:
readme 구현한 것 체크

lxxjn0 commented on .idea/.gitignore:
```diff
@@ -0,0 +1,3 @@
+# Default ignored files
```
`.idea/*` 폴더가 jetbrain IDE와 관련된 폴더더라고요.

이런 폴더 같이 버전 관리를 할 필요가 없는 폴더들은 따로 `.gitignore` 파일에 추가해서 버전 관리가 안되도록 하면 더 좋을 것 같아요 😄

JeongP replied:
아아 감사합니당 😄

JeongP committed:
view refactoring

Derek-94 commented on src/inputCarName.js:
```diff
@@ -0,0 +1,39 @@
+        validationFlagValue = false;
+        break;
+      }
+    }
```
[https://github.com/airbnb/javascript#iterators-and-generators](url) 
여기 보시면 **iterator** 관련해서 권장하는(?) 사용법이 있습니다! (11.1에 있어요.)

물론 충분히 이해가 가는 코드이고, 좋은 iterator 인데, **forEach** 구문을 쓰면 더 좋다고 명시가 되어 있더라구요!
그와 동시에 화살표 함수를 쓰는 것이 이유는 아직은 잘 모르겠는데 많이들 추천합니다. 

따라서 해당 구문을 고쳐보면
```suggestion
    inputSplit.forEach( (carName) => {
       if(carName.length > 5) {
          alert("자동차 이름의 길이를 5자 이하로 해주세요.");
          validationFlagValue = false;
       }
    }) 
```

이런식으로 수정할 수 있을것 같아요! 다만.. 감히 팁을 드리자면! **forEach** 구문은 중간에 break하기가
어렵다는 단점이 있습니다!

Derek-94 commented on index.html:
```diff
@@ -12,21 +12,22 @@ <h1>🏎️ 자동차 경주 게임</h1>
         올바른 예) east,west,south,north <br />
       </p>
-      <div class="car-game-container">
+      <div di="car-game-container">
```
오탈자가 있는 것 같아요..! di가 아니라 id로 써야할 것 같습니다..!

Derek-94 commented on src/startGame.js:
```diff
@@ -0,0 +1,75 @@
+function makeCarObj(cars) {
+  for(let i in cars) {
+    let car = new Car(cars[i]);
+    carArray.push(car);
```
이 친구도 마친가지로,
```suggestion
  cars.forEach((i) => {
     let car = new Car(i);
     carArray.push(car);
  })
```
이렇게 하시면 될 것 같습니다 👍

Derek-94 commented on src/startGame.js:
```diff
@@ -0,0 +1,75 @@
+function selectWinner() {
+  // 자동차 객체를 value 값에 따라 정렬
+  carArray.sort(function (a,b) {
+    return a.cnt > b.cnt ? -1 : a.cnt < b.cnt ? 1:0
```
sort함수에 대해서 정표님이랑 저랑 둘다 언젠간 충분한 이해가 필요한다고 생각이 듭니다 ㅜㅡㅠ 개어려워요

또한 삼항 연산자 관련해서도 많이 사용해서 익숙해지자는 의미로.. 코멘트 남겨봅니다! 😽

Derek-94 commented on src/startGame.js:
```diff
@@ -0,0 +1,75 @@
+function raceStatePrint() {
+  for(let i in carArray) {
+    let path = '';
+    if(carArray[i].cnt) {
+      for(let j=0; j<carArray[i].cnt; j++) {
+        path = path + '-';
+      }
+    } 
```
구두로 말씀드린듯이 객체가 cnt를 가지고 다닌다면 여기 indent는 해결이 가능할 것 같아요! 🚗

Derek-94 reviewed (commented):
iterator와  sort 함수 관련해서 코멘트 남겼습니다!

JeongP committed:
git ignore add

JeongP committed:
refactor: update init view

JeongP committed:
update: 오탈자 수정

JeongP committed:
refactor: update inputCarName

- 의미를 가지는 수 상수처리
- 변수와 함수명들 더 직관적으로 바꾸기
- forEach와 삼항연산자를 이용하여 깔끔하게 코드짜기

JeongP committed:
refactor: update inputCntNum

- 변수와 함수명들 더 직관적으로 바꾸기

JeongP committed:
refactor: update Race
- 파일 이름 목적에 맞게 변경(gameStart -> race)
- 함수의 depth 2이하로 낮추기.
- 함수 안에 있는 기능 단위를 바깥으로 빼내기.
- 변수와 함수들 구분하
- 개행을 로직 단위로 의미있게 사용하기.

JeongP committed:
delete: delete remark(주석)

Derek-94 commented on src/race.js:
```diff
@@ -0,0 +1,83 @@
+  return Math.floor(Math.random()*9);
+}
+
+function racing() {
+  for(let car in cars) {
+    let raceConditionNum = makeRandomNumber();
+    
+    if(raceConditionNum >= RACE_CONDITION_NUMBER) cars[car].cnt++;
+  }
+}
+
+const makePath = (car) => {
+  let path = '';
+  if(car.cnt) {
+    for(let i=0;i<car.cnt;i++) path = path + '-';
+  }
+  return path;
+}
```
함수를 선언할때는 크게 **2가지 방법**이 있다고 해요!

1. 함수 선언식
2. 함수 표현식

**함수 선언식**은, 
```javascript
function 함수명() {
  구현 로직
}
```
이런식으로 정의되는 함수입니다!

이와는 다르게 **함수 표현식**은,
```javascript
const 함수명 = function () {
  구현 로직
};
```

이런식으로 2가지로 나뉠 수가 있습니다. [https://github.com/ParkSB/javascript-style-guide#%ED%95%A8%EC%88%98-functions](url) 여기에서 간략하게 설명을 볼 수 있어요.
사실 함수 표현식을 쓰며 또 변수 참조 호출과 구분되는 이름을 쓰는 것이 호출스택을 보기에 더 좋다고는 하는데,
이름 정하는 것도 어려우니.. 이거는 더 고민해봐야 할 것 같습니다. 

함수 선언식보다 __함수 표현식__이 더 선호된다고 합니다! 
함수 선언식은 __hoisting__이 일어날 수 있는 반면, 함수 표현식은 그런 문제가 발생하지 않기 때문에, 선호 된다고 합니다. [https://joshua1988.github.io/web-development/javascript/function-expressions-vs-declarations/](url) 여기서 자세한 예시를 확인할 수 있어요.

사실 함수 표현식을 쓰며 또 변수 참조 호출과 구분되는 이름을 쓰는 것이 호출스택을 보기에 더 좋다고는 하는데,
이름 정하는 것도 어려우니.. 이거는 더 고민해봐야 할 것 같습니다..!

Derek-94 commented on src/race.js:
```diff
@@ -0,0 +1,83 @@
+
+let cars = [];
+let winnerArray = [];
+const RACE_CONDITION_NUMBER = 4;
+const COMMA = ', ';
+const COMMA_BLANK = 2;
```
이게 맞는건지는 모르겠는데, 일단 참고만 하는게 좋을 것 같아서 남겨봅니다 :)

이런 고정된 상수 변수를 모아두어서 모듈화 시키는것도 좋을것 같아요.

```suggestion
const constantValues = {
  RACE_CONDITION_NUMBER = 4,
  COMMA = ', ',
  COMMA_BLANK = 2
}
```
이렇게 만들고, 또 `alert` 메세지들도 모으는 게 좋을 것 같아요.

```javascript
const message = {
CAR_NAME_5_LETTERS_LIMIT : '자동차 이름의 길이를 5자 이하로 해주세요.',
...
}
```

이런 고정된 친구들을 모아서 다른폴더에 넣고, 이를 다시 `import` 하는게 어떨까요?
```javascript
import {conatntValues, message} from "../constants.js";

...

if(carNameLength > 5) {
  alert(message.CAR_NAME_5_LETTERS_LIMIT);
}
```

이런식으로 정리해도 보기에 깔끔할 것 같기도 합니당! 

그냥 제 의견이긴합니다.. ㅎㅎ

JeongP replied:
와우...! 이렇게 하면 상수나 msg를 한 번에 관리할 수 있겠네요 
감사합니닷 :)

Derek-94 reviewed (approved):
음.. 크게 두가지만 남겨두었어욥 

이 racing car 를 다시 다 뜯어고친다기 보다는 추후에 작성될 함수들이나 
상수들에 대해서 반영되면 좋을것 같은.. 점을 리뷰했습니당:)

JeongP committed:
refactor: 상수 따로 파일로 처리
